File: src/gui/dialogs/preferences_dialog.py
"""
Preferences Dialog
"""

from PyQt6.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QLabel, 
    QComboBox, QPushButton, QCheckBox, QTabWidget,
    QWidget, QFormLayout, QSpinBox, QLineEdit, QFileDialog
)
from PyQt6.QtCore import Qt
from utils.config import Config

class PreferencesDialog(QDialog):
    """Application preferences dialog"""
    
    def __init__(self, config: Config, parent=None):
        super().__init__(parent)
        self.config = config
        self.setWindowTitle("Preferences")
        self.resize(600, 400)
        
        layout = QVBoxLayout(self)
        
        # Tabs
        self.tabs = QTabWidget()
        self.tabs.addTab(self._create_general_tab(), "General")
        self.tabs.addTab(self._create_paths_tab(), "Paths")
        self.tabs.addTab(self._create_ai_tab(), "AI Models")
        
        # Model Manager tab
        from gui.dialogs.model_manager_widget import ModelManagerWidget
        self.model_manager = ModelManagerWidget()
        self.tabs.addTab(self.model_manager, "📦 Model Manager")
        
        layout.addWidget(self.tabs)
        
        # Buttons
        btn_layout = QHBoxLayout()
        btn_cancel = QPushButton("Cancel")
        btn_cancel.clicked.connect(self.reject)
        
        btn_save = QPushButton("Save")
        btn_save.clicked.connect(self._save_preferences)
        btn_save.setStyleSheet("background-color: #4CAF50; color: white; font-weight: bold;")
        
        btn_layout.addStretch()
        btn_layout.addWidget(btn_cancel)
        btn_layout.addWidget(btn_save)
        
        layout.addLayout(btn_layout)
        
    def _create_general_tab(self):
        widget = QWidget()
        form = QFormLayout(widget)
        
        # Theme
        self.combo_theme = QComboBox()
        self.combo_theme.addItems(["dark", "light", "system"])
        self.combo_theme.setCurrentText(self.config.get('gui.theme', 'dark'))
        form.addRow("Theme:", self.combo_theme)
        
        # Auto-save
        self.chk_autosave = QCheckBox("Enable Auto-save")
        self.chk_autosave.setChecked(self.config.get('gui.autosave', True))
        form.addRow("Auto-save:", self.chk_autosave)
        
        self.spin_interval = QSpinBox()
        self.spin_interval.setRange(60, 3600)
        self.spin_interval.setSuffix(" seconds")
        self.spin_interval.setValue(self.config.get('gui.autosave_interval', 300))
        form.addRow("Interval:", self.spin_interval)
        
        return widget
        
    def _create_paths_tab(self):
        widget = QWidget()
        form = QFormLayout(widget)
        
        # Output Dir
        self.txt_output = QLineEdit(self.config.get('processing.output_dir', 'output/'))
        btn_output = QPushButton("Browse...")
        btn_output.clicked.connect(lambda: self._browse_dir(self.txt_output))
        
        out_layout = QHBoxLayout()
        out_layout.addWidget(self.txt_output)
        out_layout.addWidget(btn_output)
        form.addRow("Output Directory:", out_layout)
        
        # Temp Dir
        self.txt_temp = QLineEdit(self.config.get('processing.temp_dir', 'temp/'))
        btn_temp = QPushButton("Browse...")
        btn_temp.clicked.connect(lambda: self._browse_dir(self.txt_temp))
        
        temp_layout = QHBoxLayout()
        temp_layout.addWidget(self.txt_temp)
        temp_layout.addWidget(btn_temp)
        form.addRow("Temp Directory:", temp_layout)
        
        return widget
        
    def _create_ai_tab(self):
        widget = QWidget()
        form = QFormLayout(widget)
        
        # UVR
        form.addRow(QLabel("<b>Vocal Separation (UVR)</b>"))
        self.chk_gpu = QCheckBox("Use GPU Acceleration")
        self.chk_gpu.setChecked(self.config.get('uvr.use_gpu', True))
        form.addRow("Use GPU:", self.chk_gpu)
        
        # Whisper
        form.addRow(QLabel("<b>Lyrics Transcription (Whisper)</b>"))
        self.combo_whisper = QComboBox()
        
        # Auto-detect available Whisper models
        from pathlib import Path
        whisper_dir = Path("models/whisper")
        available_models = []
        
        if whisper_dir.exists():
            # Find all .pt files
            for model_file in whisper_dir.glob("*.pt"):
                model_name = model_file.stem  # e.g., "small", "medium"
                available_models.append(model_name)
        
        # Add standard models (in order of quality)
        standard_models = ["tiny", "base", "small", "medium", "large", "large-v2", "large-v3"]
        
        for model in standard_models:
            if model in available_models:
                self.combo_whisper.addItem(f"{model} ✓ (Downloaded)", model)
            else:
                self.combo_whisper.addItem(f"{model} (Will download)", model)
        
        # Set current model
        current_model = self.config.get('lyrics.whisper_model', 'small')
        index = self.combo_whisper.findData(current_model)
        if index >= 0:
            self.combo_whisper.setCurrentIndex(index)
        
        form.addRow("Model Size:", self.combo_whisper)
        
        # Add info label
        info_label = QLabel(
            "💡 Downloaded models have a ✓ mark.\n"
            "Larger models = better accuracy but slower."
        )
        info_label.setWordWrap(True)
        info_label.setStyleSheet("color: #666; font-size: 9pt; font-style: italic;")
        form.addRow("", info_label)
        
        return widget
        
    def _browse_dir(self, line_edit):
        path = QFileDialog.getExistingDirectory(self, "Select Directory", line_edit.text())
        if path:
            line_edit.setText(path)
            
    def _save_preferences(self):
        # Save values to config
        self.config.set('gui.theme', self.combo_theme.currentText())
        self.config.set('gui.autosave', self.chk_autosave.isChecked())
        self.config.set('gui.autosave_interval', self.spin_interval.value())
        
        self.config.set('processing.output_dir', self.txt_output.text())
        self.config.set('processing.temp_dir', self.txt_temp.text())
        
        self.config.set('uvr.use_gpu', self.chk_gpu.isChecked())
        
        # Get Whisper model from userData (not text, since text has ✓ mark)
        whisper_model = self.combo_whisper.currentData()
        if not whisper_model:  # Fallback if no userData
            whisper_model = self.combo_whisper.currentText().split()[0]
        self.config.set('lyrics.whisper_model', whisper_model)
        
        # Save to file
        self.config.save()
        self.accept()

File: cpp/src/gui/dialogs/preferences_dialog.cpp
/*
 * NC-KTV GUI — Preferences Dialog Implementation
 */

#include "preferences_dialog.h"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFormLayout>
#include <QGroupBox>
#include <QFileDialog>
#include <QCoreApplication>

namespace ncktv {

PreferencesDialog::PreferencesDialog(ConfigManager* config, QWidget* parent)
    : QDialog(parent)
    , m_config(config)
{
    setWindowTitle("Preferences");
    setMinimumSize(560, 460);
    setModal(true);
    setupUi();
    applyTheme();
}

void PreferencesDialog::setupUi() {
    auto* mainLayout = new QVBoxLayout(this);
    mainLayout->setSpacing(14);
    mainLayout->setContentsMargins(20, 16, 20, 16);

    auto* titleLabel = new QLabel("⚙  Preferences", this);
    titleLabel->setStyleSheet("font-size:18px; font-weight:bold; color:#e0e0e0; border:none;");
    mainLayout->addWidget(titleLabel);

    // ── Tab Widget ───────────────────────────────────────────────────────
    m_tabs = new QTabWidget(this);
    m_tabs->addTab(createGeneralTab(), "General");
    m_tabs->addTab(createAudioTab(), "Audio");
    m_tabs->addTab(createAiTab(), "AI Models");
    m_tabs->addTab(createPathsTab(), "Paths");
    mainLayout->addWidget(m_tabs, 1);

    // ── Buttons ──────────────────────────────────────────────────────────
    auto* btnLayout = new QHBoxLayout();
    btnLayout->addStretch();
    m_cancelBtn = new QPushButton("Cancel", this);
    m_cancelBtn->setFixedWidth(100);
    btnLayout->addWidget(m_cancelBtn);
    m_okBtn = new QPushButton("Save", this);
    m_okBtn->setFixedWidth(100);
    m_okBtn->setDefault(true);
    m_okBtn->setStyleSheet("background:#4f46e5; color:white; font-weight:bold; border:none; padding:8px 16px; border-radius:4px;");
    btnLayout->addWidget(m_okBtn);
    mainLayout->addLayout(btnLayout);

    connect(m_okBtn, &QPushButton::clicked, this, &PreferencesDialog::onAccepted);
    connect(m_cancelBtn, &QPushButton::clicked, this, &QDialog::reject);
}

QWidget* PreferencesDialog::createGeneralTab() {
    auto* widget = new QWidget(this);
    auto* form = new QFormLayout(widget);
    form->setSpacing(12);
    form->setContentsMargins(16, 16, 16, 16);

    m_startModeCombo = new QComboBox(this);
    m_startModeCombo->addItems({"Wizard", "Editor"});
    form->addRow("Start Mode:", m_startModeCombo);

    m_graphicsApiCombo = new QComboBox(this);
    m_graphicsApiCombo->addItems({"Default", "Vulkan (High Perf)", "DirectX 11", "OpenGL", "Software"});
    QString currentApi = m_config ? m_config->get<QString>("gui.graphics_api", "Default") : "Default";
    m_graphicsApiCombo->setCurrentText(currentApi);
    form->addRow("Graphics Engine:", m_graphicsApiCombo);
    
    auto* graphicsNote = new QLabel("Note: Changing the graphics engine requires an application restart.", this);
    graphicsNote->setStyleSheet("color:#8a8a8a; font-size:10px; font-style:italic; border:none;");
    form->addRow("", graphicsNote);

    m_autoSaveCheck = new QCheckBox("Enable auto-save", this);
    m_autoSaveCheck->setChecked(true);
    form->addRow("", m_autoSaveCheck);

    m_autoSaveInterval = new QSpinBox(this);
    m_autoSaveInterval->setRange(30, 600);
    m_autoSaveInterval->setValue(120);
    m_autoSaveInterval->setSuffix(" seconds");
    form->addRow("Auto-save interval:", m_autoSaveInterval);

    return widget;
}

QWidget* PreferencesDialog::createAudioTab() {
    auto* widget = new QWidget(this);
    auto* form = new QFormLayout(widget);
    form->setSpacing(12);
    form->setContentsMargins(16, 16, 16, 16);

    m_sampleRateCombo = new QComboBox(this);
    m_sampleRateCombo->addItems({"44100 Hz", "48000 Hz", "96000 Hz"});
    m_sampleRateCombo->setCurrentIndex(1);
    form->addRow("Sample Rate:", m_sampleRateCombo);

    m_gpuCheck = new QCheckBox("Use GPU for vocal separation (CUDA/DirectML)", this);
    m_gpuCheck->setChecked(true);
    form->addRow("", m_gpuCheck);

    return widget;
}

QWidget* PreferencesDialog::createAiTab() {
    auto* widget = new QWidget(this);
    auto* form = new QFormLayout(widget);
    form->setSpacing(12);
    form->setContentsMargins(16, 16, 16, 16);

    // ── Transcription Engine ────────────────────────────────────────────
    m_transcriptionEngineCombo = new QComboBox(this);
    m_transcriptionEngineCombo->addItems({"WhisperX (Recommended)", "Whisper (Legacy)"});
    QString currentEngine = m_config ? m_config->get<QString>("ai.transcription_engine", "whisperx") : "whisperx";
    if (currentEngine == "whisper") {
        m_transcriptionEngineCombo->setCurrentIndex(1);
    } else {
        m_transcriptionEngineCombo->setCurrentIndex(0);
    }
    form->addRow("Transcription Engine:", m_transcriptionEngineCombo);

    auto* engineNote = new QLabel(
        "<b>WhisperX</b>: Uses Wav2Vec2 forced alignment for ±30ms word precision. "
        "Ideal for karaoke wipe effects.<br>"
        "<b>Whisper</b>: Legacy mode with attention-based timestamps (~±500ms).",
        this);
    engineNote->setWordWrap(true);
    engineNote->setStyleSheet("color:#64748b; font-size:10px; border:none; margin-bottom:6px;");
    form->addRow("", engineNote);

    // ── Whisper Model ───────────────────────────────────────────────────
    m_defaultModelCombo = new QComboBox(this);
    m_defaultModelCombo->addItems({"base", "small", "medium", "large-v2", "large-v3", "turbo"});
    QString currentWhisper = m_config ? m_config->get<QString>("ai.whisper_model", "medium") : "medium";
    m_defaultModelCombo->setCurrentText(currentWhisper);
    form->addRow("Default Whisper Model:", m_defaultModelCombo);

    // ── UVR Model ───────────────────────────────────────────────────────
    m_defaultUvrModelCombo = new QComboBox(this);
    QStringList uvrPaths = {
        QCoreApplication::applicationDirPath() + "/models/uvr",
        QDir::currentPath() + "/models/uvr"
    };
    QStringList uvrModels;
    for (const auto& path : uvrPaths) {
        QDir dir(path);
        if (dir.exists()) {
            uvrModels << dir.entryList({"*.onnx", "*.pth"}, QDir::Files);
        }
    }
    uvrModels.removeDuplicates();
    if (uvrModels.isEmpty()) uvrModels << "UVR_MDXNET_KARA_2.onnx";
    m_defaultUvrModelCombo->addItems(uvrModels);
    
    QString currentUvr = m_config ? m_config->get<QString>("ai.uvr_model", "UVR_MDXNET_KARA_2.onnx") : "UVR_MDXNET_KARA_2.onnx";
    m_defaultUvrModelCombo->setCurrentText(currentUvr);
    form->addRow("Default UVR Model:", m_defaultUvrModelCombo);

    // ── Language ────────────────────────────────────────────────────────
    m_defaultLanguageCombo = new QComboBox(this);
    m_defaultLanguageCombo->addItems({"Auto", "en", "ms", "id", "ja", "ko", "zh"});
    QString currentLang = m_config ? m_config->get<QString>("ai.language", "Auto") : "Auto";
    m_defaultLanguageCombo->setCurrentText(currentLang);
    form->addRow("Default Language:", m_defaultLanguageCombo);

    auto* noteLabel = new QLabel("Larger models produce better results but require more RAM and processing time.", this);
    noteLabel->setWordWrap(true);
    noteLabel->setStyleSheet("color:#888; font-size:11px; border:none;");
    form->addRow("", noteLabel);

    return widget;
}

QWidget* PreferencesDialog::createPathsTab() {
    auto* widget = new QWidget(this);
    auto* form = new QFormLayout(widget);
    form->setSpacing(12);
    form->setContentsMargins(16, 16, 16, 16);

    auto* ffmpegLayout = new QHBoxLayout();
    m_ffmpegPathEdit = new QLineEdit(this);
    m_ffmpegPathEdit->setPlaceholderText("Auto-detected");
    auto* ffmpegBtn = new QPushButton("Browse...", this);
    ffmpegBtn->setFixedWidth(80);
    ffmpegLayout->addWidget(m_ffmpegPathEdit, 1);
    ffmpegLayout->addWidget(ffmpegBtn);
    form->addRow("FFmpeg Path:", ffmpegLayout);
    connect(ffmpegBtn, &QPushButton::clicked, this, &PreferencesDialog::browseFfmpegPath);

    auto* modelLayout = new QHBoxLayout();
    m_modelPathEdit = new QLineEdit(this);
    m_modelPathEdit->setPlaceholderText("models/whisper");
    auto* modelBtn = new QPushButton("Browse...", this);
    modelBtn->setFixedWidth(80);
    modelLayout->addWidget(m_modelPathEdit, 1);
    modelLayout->addWidget(modelBtn);
    form->addRow("Model Directory:", modelLayout);
    connect(modelBtn, &QPushButton::clicked, this, &PreferencesDialog::browseModelPath);

    return widget;
}

void PreferencesDialog::applyTheme() {
    setStyleSheet(R"(
        QDialog { background:#1e1e1e; color:#d4d4d4; }
        QTabWidget::pane { border:1px solid #3f3f46; border-radius:4px; background:#1e1e1e; }
        QTabBar::tab { background:#252526; color:#ccc; padding:8px 16px; border:1px solid #3f3f46; border-bottom:none; border-top-left-radius:4px; border-top-right-radius:4px; }
        QTabBar::tab:selected { background:#1e1e1e; color:#fff; }
        QTabBar::tab:hover { background:#333; }
        QLineEdit, QComboBox, QSpinBox { background:#2d2d30; border:1px solid #3f3f46; border-radius:4px; padding:6px 8px; color:#fff; }
        QComboBox::drop-down { border:none; }
        QPushButton { background:#333; border:1px solid #555; border-radius:4px; padding:6px 14px; color:#ccc; }
        QPushButton:hover { background:#404040; }
        QCheckBox { color:#ccc; }
        QLabel { color:#ccc; border:none; }
    )");
}

void PreferencesDialog::browseFfmpegPath() {
    QString dir = QFileDialog::getExistingDirectory(this, "Select FFmpeg Directory");
    if (!dir.isEmpty()) m_ffmpegPathEdit->setText(dir);
}

void PreferencesDialog::browseModelPath() {
    QString dir = QFileDialog::getExistingDirectory(this, "Select Model Directory");
    if (!dir.isEmpty()) m_modelPathEdit->setText(dir);
}

void PreferencesDialog::onAccepted() {
    if (m_config) {
        m_config->set<QString>("gui.start_mode", m_startModeCombo->currentText().toLower());
        m_config->set<QString>("gui.graphics_api", m_graphicsApiCombo->currentText());
        // Transcription engine: index 0 = whisperx, index 1 = whisper
        m_config->set<QString>("ai.transcription_engine",
            m_transcriptionEngineCombo->currentIndex() == 0 ? "whisperx" : "whisper");
        m_config->set<QString>("ai.whisper_model", m_defaultModelCombo->currentText());
        m_config->set<QString>("ai.uvr_model", m_defaultUvrModelCombo->currentText());
        m_config->set<QString>("ai.language", m_defaultLanguageCombo->currentText());
        m_config->save();
    }
    accept();
}

} // namespace ncktv
